### DISCUSSION TITLE
classes/class_ninepatchrect

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#164] [User-contributed notes]
# classes/class_ninepatchrect

Inherits: Control< CanvasItem< Node< Object A control that displays a texture by keeping its corners intact, but tiling its edges and center. Description: Also known as 9-slice panels, NinePatchRec...

https://docs.godotengine.org/en/stable/classes/class_ninepatchrect.html

**Comment by devmark404** [comment]
I have created a tutorial that shows how to use it and where its name comes from along with some recommendations. It is in Spanish, but has subtitles in other languages

Article and Video
https://godot.land/godot-nodo-ninepatchrect
Video
https://youtu.be/Icw3s8luiXc

**Comment by chuckwolber** [comment]
I developed this to help visualize the behavior a bit easier.

You start out with a texture divided into nine sections:

```
+---+---+---+
| 1 | 2 | 3 |
+---+---+---+
| 4 | 5 | 6 |
+---+---+---+
| 7 | 8 | 9 |
+---+---+---+
```

If axis stretch mode is `AXIS_STRETCH_MODE_STRETCH`, stretching horizontally and vertically will get you:

```
+---+-----------+---+
| 1 |     2     | 3 |
+---+-----------+---+
|   |           |   |
| 4 |     5     | 6 |
|   |           |   |
+---+-----------+---+
| 7 |     8     | 9 |
+---+-----------+---+
```

If axis stretch mode is `AXIS_STRETCH_MODE_TILE`, stretching horizontally and vertically will get you:

```
+---+---------------+---+
| 1 | 2 | 2 | 2 | 2 | 3 |
+---+---------------+---+
| 4 | 5 | 5 | 5 | 5 | 6 |
| 4 | 5 | 5 | 5 | 5 | 6 |
| 4 | 5 | 5 | 5 | 5 | 6 |
+---+---------------+---+
| 7 | 8 | 8 | 8 | 8 | 9 |
+---+---------------+---+
```

**Reply by stevencasteel** [reply]
Reminds me of NES Mega Man's password system!

Thanks for the visualization, that helped a ton.

Gemini 2.5 led me here. Here's its advice from an audit of my project:

For UI Layout: NinePatchRect

What you're doing now:
Using procedurally drawn ColorRects for your StyledMenuItem. This is great for a minimalist style and gives you total control.

What NinePatchRect provides:
A way to create scalable UI panels from a small texture without distortion. It understands which parts of an image are corners (don't stretch), which are edges (stretch on one axis), and which is the center (stretch on both axes).

Why it's like the Theme resource (powerful Godot thing): 
It's a built-in solution for a universal UI problem: creating buttons and panels that can be resized to fit any text length without ugly stretching.

How you would use it:
If you ever wanted to give your buttons a textured border (like a metallic frame or a beveled edge), you would draw one small NinePatchRect texture and use it for every button, no matter the size.

**Reply by stevencasteel** [reply]
Gemini 3.1's audit of the comment:

Gemini 2.5 perfectly explains the concept of 9-slice scaling and why it's so useful for creating scalable UI without texture distortion. 

However, it misses a crucial Godot best practice: to style standard UI elements like buttons, you should actually use a StyleBoxTexture inside the Theme system rather than manually adding NinePatchRect nodes everywhere.

**Comment by No1Idle** [comment]
What does the AXIS_STRETCH_MODE_TILE do when the center part is less than one repetion?, Like if the center part is 100 pixels and you have space just for 50 pixels? Are only 50 pixels of the middle displayed or the 100 pixels will be "squeezed" in 50?

**Reply by Calinou** [reply]
Assuming a 64×64 texture with 4-pixel borders on all edges:

<img width="255" height="246" alt="image" src="https://github.com/user-attachments/assets/b666db70-ff76-49c8-970f-e9908e07a994" />

It'll look like this when the node's size is not a multiple of 64×64:

Stretch | Tile | Tile Fit
-|-|-
<img width="785" height="470" alt="image" src="https://github.com/user-attachments/assets/c3ff609c-e185-45f0-92ac-1869f59756bd" /> | <img width="785" height="470" alt="image" src="https://github.com/user-attachments/assets/9be726ad-c224-4618-9715-7fe675286c5e" /> | <img width="785" height="470" alt="image" src="https://github.com/user-attachments/assets/08998e47-0476-41a8-84f0-8286f4d41071" />

This also applies if the node is smaller than 64×64; the tiled texture will be cut off at the bottom/right sides.
